kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/expressions/DefaultExpressionUtils.java

```diff
@@ -87,46 +101,87 @@ public boolean getBoolean(int rowId) {
+     * comparator ex. vectorComparator (ie. ==, >=, <=......) for left and right
+     * column vector according to the natural ordering of numbers
      * <p>
      * Only primitive data types are supported.
      */
-    static int[] compare(ColumnVector left, ColumnVector right) {
+    static ColumnVector comparatorVector(
+            ColumnVector left,
+            ColumnVector right,
+            VectorComparator vectorComparator) {
         checkArgument(
-            left.getSize() == right.getSize(),
-            "Left and right operand have different vector sizes.");
-        DataType dataType = left.getDataType();
+                left.getSize() == right.getSize(),
+                "Left and right operand have different vector sizes.");
+        return new ColumnVector() {
 
-        int numRows = left.getSize();
-        int[] result = new int[numRows];
-        if (dataType instanceof BooleanType) {
-            compareBoolean(left, right, result);
-        } else if (dataType instanceof ByteType) {
-            compareByte(left, right, result);
-        } else if (dataType instanceof ShortType) {
-            compareShort(left, right, result);
-        } else if (dataType instanceof IntegerType || dataType instanceof DateType) {
-            compareInt(left, right, result);
-        } else if (dataType instanceof LongType ||
-                dataType instanceof TimestampType ||
-                dataType instanceof TimestampNTZType) {
-            compareLong(left, right, result);
-        } else if (dataType instanceof FloatType) {
-            compareFloat(left, right, result);
-        } else if (dataType instanceof DoubleType) {
-            compareDouble(left, right, result);
-        } else if (dataType instanceof DecimalType) {
-            compareDecimal(left, right, result);
-        } else if (dataType instanceof StringType) {
-            compareString(left, right, result);
-        } else if (dataType instanceof BinaryType) {
-            compareBinary(left, right, result);
-        } else {
-            throw new UnsupportedOperationException(dataType + " can not be compared.");
-        }
-        return result;
+            @Override
+            public DataType getDataType() {
+                return BooleanType.BOOLEAN;
+            }
+
+            @Override
+            public void close() {
+                Utils.closeCloseables(left, right);
+            }
+
+            @Override
+            public int getSize() {
+                return left.getSize();
+            }
+
+            @Override
+            public boolean isNullAt(int rowId) {
+                return left.isNullAt(rowId) || right.isNullAt(rowId);
+            }
+
+            @Override
+            public boolean getBoolean(int rowId) {
+                if (isNullAt(rowId)) {
+                    return false;
+                }
+                DataType dataType = left.getDataType();
```

Comment:
The `VectorComparator` interface looks good.

Only this is that we now evaluate the this long `if` statement for every row. Instead of we create a comparator operator for each type (e.g. boolean or short) we can avoid the `if` statement cost.


```
class ValueComparator {
   ValueComparator(ColumnVector v1, ColumnVector v2); 

    boolean getCompareResult(int rowId);
}

class ByteValueComparator {
   ColumnVector v1;
   ColumnVector v2;

   ByteValueComparator(ColumnVector v1, ColumnVector v2) {
      this.v1 = v1;
      this.v2 = v2;
   }
   boolean getCompareResult(int rowId) {
      // null check 
     // and then compare value for non-nulls
       Byte.compare(left.getByte(rowId), right.getByte(rowId)))
   }
}

Reply:
Hi @vkorukanti  good catch, I didn't thought about that and should move the long if statement outside of anonymous class. I have fixed the suggestion, please let me know if this is what you are looking for
thanks